plugins/rbac-backend/src/service/role-manager.ts

```diff
@@ -1,40 +1,149 @@
+  async hasLink(
+    name1: string,
+    name2: string,
+    ...domain: string[]
   ): Promise<boolean> {
-    throw new Error('Method not implemented.');
+    if (domain.length > 0) {
+      throw new Error('domain argument is not supported.');
+    }
+
+    if (name1 === name2) {
+      return true;
+    }
+
+    // name1 is always user in our case.
+    // name2 is user or group.
+    // user(name1) couldn't inherit user(name2).
+    // We can use this fact for optimization.
+    const { kind } = parseEntityRef(name2);
+    if (kind.toLocaleLowerCase() === 'user') {
+      return false;
+    }
+
+    // optimization: don't make request if name2 is user...
+    const role = await this.findAncestorGroup(
+      [name1],
+      name2,
+      'relations.hasMember',
+    );
+    return !!role;
   }
+
+  /**
+   * syncedHasLink determines whether role: name1 inherits role: name2.
+   * domain is a prefix to the roles.
+   */
   syncedHasLink?(
     _name1: string,
     _name2: string,
     ..._domain: string[]
   ): boolean {
-    throw new Error('Method not implemented.');
+    throw new Error('Method "syncedHasLink" not implemented.');
   }
-  getRoles(_name: string, ..._domain: string[]): Promise<string[]> {
-    throw new Error('Method not implemented.');
+
+  /**
+   * getRoles gets the roles that a subject inherits.
+   * domain is a prefix to the roles.
+   */
+  async getRoles(_name: string, ..._domain: string[]): Promise<string[]> {
+    throw new Error('Method "getRoles" not implemented.');
+  }
+
+  /**
+   * getUsers gets the users that inherits a subject.
+   * domain is an unreferenced parameter here, may be used in other implementations.
+   */
+  async getUsers(_name: string, ..._domain: string[]): Promise<string[]> {
+    throw new Error('Method "getUsers" not implemented.');
   }
-  getUsers(_name: string, ..._domain: string[]): Promise<string[]> {
-    throw new Error('Method not implemented.');
+
+  /**
+   * printRoles prints all the roles to log.
+   */
+  async printRoles(): Promise<void> {
+    // do nothing
   }
-  printRoles(): Promise<void> {
-    throw new Error('Method not implemented.');
+
+  private async findAncestorGroup(
+    entityRefs: string[],
+    groupToSearch: string,
+    fr: FilterRelations,
+  ): Promise<Entity | undefined> {
+    const { items } = await this.catalogApi.getEntities({
+      filter: {
+        kind: 'Group',
+        [fr]: Array.from(entityRefs),
+      },
+      // Save traffic with only required information for us
+      fields: ['metadata.name', 'kind', 'metadata.namespace', 'spec.parent'],
+    });
+
+    const groupsRefs = new Set<string>();
+    for (const item of items) {
+      const groupRef = `group:default/${item.metadata.name.toLocaleLowerCase()}`;
+      if (groupRef === groupToSearch) {
+        return item;
+      }
+      if (item.spec?.parent) {
+        groupsRefs.add(groupRef);
+      }
+    }
+
+    if (groupsRefs.size > 0) {
+      return await this.findAncestorGroup(
```

Comment:
What happens if we have a cyclic group relationship? For instance, user A -> group ZZ -> group YY -> group XX -> group ZZ ?

Reply:
https://github.com/casbin/casbin/issues/338 casbin resolves such situation with DefaultRoleManager with constructor option maxDepthLimit. We can implement the same approach and make it configurable via application config. With this option users could even disable group inheritance if they wants to have groups permission with one group level. Or users could disable group feature at all, but I'm not sure if that a good option in real world.

Reply:
Enother possible option: path tracking. Path Tracking: RBAC systems can keep track of the roles or groups that have already been visited during permission evaluation. If a role or group is encountered again in the evaluation process, it can be flagged as a cycle, and the system can take appropriate action to handle the cycle, such as denying access.

Reply:
I like first option: maxDepthLimit

Reply:
Also we can control performance with maxDepthLimit option.